Assistant: New application similar to macOS Spotlight

SpencerCDixon opened the pull request:
Hey folks,

I've created a new GUI application similar to macOS Spotlight. I'm a big user of Spotlight/Alfred/launcher applications and felt like this was a void that could be filled for Serenity.

The app is called `Assistant` because it's job is to assist you in system tasks. The first two pieces of functionality it has are:

1. Opening applications quickly
2. Doing math calculations and copying the results to your clipboard

A new global hotkey was added to trigger the opening of the Assistant (`Super+Space`).

It's likely this PR is not ready to merge (still missing icons and may still be bugs I haven't discovered) but I wanted to put it out for feedback and see what everyone thinks.

Here is a quick demo:

https://user-images.githubusercontent.com/7471018/123526307-06fb1c00-d6a5-11eb-9bc7-fd866bda750d.mp4

---

Again, very open to feedback and would love to get anything fixed up that needs to be done to get this merged so I can slowly add more functionality over time. Along those lines, the app was structured in a way to make it fairly easy for people to add new `Assistant::Provider`'s of their own so if anyone feels passionately about adding some functionality it should be fairly straightforward.

linusg commented:
> Assistant: Fix clang-format issues

Please don't do that. Any fixups must be squashed into the appropriate previous commit(s) instead.

Generally I think some of these commits don't need to exist on their own:

> Assistant: Remove unneeded Widget::update()

If it's not needed, don't add it in the first place? Should be squashed into the initial Assistant commit where it was added needlessly.

Same for these:

> Assistant: Include the new serenity_component for building Assistant
> Assistant: Remove cached JS::VM since it was leading to crashes
> Assistant: Set proper pledges needed for the app to run

They're all fixes, but we don't need to keep the original history how the app was developed :)

linusg reviewed (commented):
Please use east const in all new code - see https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#const-placement, https://mariusbancila.ro/blog/2018/11/23/join-the-east-const-revolution/

alimpfard commented on Userland/Applications/Assistant/main.cpp:
```diff
@@ -206,29 +207,91 @@ class CalculatorProvider : public Provider {
-    // TODO: Cache results
-    Vector<NonnullRefPtr<Result>> m_results;
+    Threading::Lock m_lock;
+    HashMap<String, Vector<AK::NonnullRefPtr<Result>>> m_result_cache;
```
`AK::` is not necessary unless in really rare cases for disambiguation.
all public AK classes/structs are imported into the global namespace.

alimpfard commented on Userland/Applications/Assistant/FuzzyMatch.cpp:
```diff
@@ -0,0 +1,170 @@
+
+namespace Assistant {
+
+static bool fuzzy_match_recursive(const char* pattern, const char* str, int& outScore, const char* strBegin,
+    uint8_t const* srcMatches, uint8_t* newMatches, int maxMatches, int nextMatch,
+    int& recursionCount, int recursionLimit);
```
Please don't use camelCase for variable names.
also, this function is pretty damn chunky.
also, please prefer `StringView` over `const char*` where possible.
also, most of these `int`'s are actually indices/counts, so please use `size_t` for those.
also, please prefer `uN` over `uintN_t` (e.g. `u8` here).
also, `matches` + `maxMatches` sounds like it could be a `Span<u8>`.

I'm not sure what the algorithm is, but it looks to me like you're using `strBegin + some_scan_index` as the string, so perhaps switch to a `StringView` plus an index instead?

SpencerCDixon replied:
So yeah wasn't sure what to do about this. The fuzzy algorithm is taken from [here](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), you can see my note [here](https://github.com/SerenityOS/serenity/pull/8266/files/eda5de6c68d32d433bc7a2acd61ab22d6cac2cb4..6610d1932826a211cae869fad6b26a1e529fee75#diff-5816677caa62899602186d3731f71db96d6fc232f6dbb83fe9f3cd2e591e660fR13)

I intentionally preserved the code from the article but I'm happy to port it into a 'SerenityOS' style if that is preferred. Wasn't sure if there was value in leaving it the way it was originally written or not.

alimpfard replied:
Importing code isn't really something we do here, it's definitely alright to look at other codebases and cite them as a source, but just copy-paste isn't really an option.

SpencerCDixon replied:
Yup fair enough, chatted with @awesomekling about that. Just wrote my own version of the fuzzy matching algorithm using AK datastructures/idioms

alimpfard commented on Userland/Applications/Assistant/FuzzyMatch.cpp:
```diff
@@ -0,0 +1,170 @@
+                // Camel case
+                char neighbor = strBegin[currIdx - 1];
+                char curr = strBegin[currIdx];
+                if (::islower(neighbor) && ::isupper(curr))
```
There's no other function named `islower` or `isupper` in this scope, I'd say the `::` is redundant.

alimpfard commented on Userland/Applications/Assistant/FuzzyMatch.cpp:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+}
```
This is repeated quite a few times, and I'm sure the CI will pick up on it, but please leave an empty line at the end of files.

alimpfard reviewed (changes requested):
A very, very surface-level review, but we've gotta start somewhere :^)
- Please observe the contribution guidelines and the code style document (again)
- As Linus also mentioned, a lot of these commits can be reordered and squashed into what they fix (specially formatting/style changes); more specifically, please avoid trivial commit dependencies that can be resolved by reordering (e.g. A -> B -> C => B -> A+C)

SpencerCDixon commented:
@linusg / @alimpfard looks like the CI may have timed out but I believe I got all linting issues fixed. Also, refactored to `east` const, fixed the Fuzzy find impl, and cleaned up the git history.

Think this should be ready to merge unless there are some other things that y'all want changed/fixed!

awesomekling commented on Userland/Libraries/LibGUI/Widget.cpp:
```diff
@@ -891,16 +891,19 @@ Vector<Widget&> Widget::child_widgets() const
 void Widget::set_palette(const Palette& palette)
 {
     m_palette = palette.impl();
+    update();
```
These changes are already in master: 5b79d53c893315f489c352b627071f5bd4c8dd36

SpencerCDixon replied:
Removed commit

awesomekling commented on Userland/Libraries/LibGUI/Event.h:
```diff
@@ -102,6 +103,14 @@ class WMSuperKeyPressedEvent : public WMEvent {
     }
 };
 
```
The commit message refers to `Taskbar.app` and `Assistant.app` but `.app` suffixes are not a thing on SerenityOS. :^)

SpencerCDixon replied:
Fixed!

awesomekling commented:
Very cool! This is such a cool idea and I think it will fit great with the system.

There are various little things that can be made more efficient, but let's merge the app first and not worry too much about performance until it's an actual issue :)

Will be interesting to see what kind of providers we can add to this!

Mandar12 commented on Userland/Libraries/LibGUI/Event.h:
```diff
@@ -63,6 +63,7 @@ class Event : public Core::Event {
         WM_WindowIconBitmapChanged,
         WM_AppletAreaSizeChanged,
         WM_SuperKeyPressed,
+        WM_SuperSpaceKeyPressed,
```
@SpencerCDixon Alfred provides Hotkey preference.
Assistant could also have configuration to let user choose specific keys.

This would require Assistant to register those keys with Taskbar as trigger key.

Also a persistent preference/configuration can be used for more use cases

SpencerCDixon replied:
Yeah that would be really cool to add one day! Talked to Kling about that initially and decided to keep it simple with hard coded action to start with idea of adding a generalized hotkey system in the future potentially.
